File: ios/AtlasCore/Sources/AtlasCore/Keys/Enclave.swift
import Foundation

/// Secure Enclave biometric-bound key release (§0.3, §7). Mirrors
/// `backend/atlas/keys/enclave.py`.
///
/// Device-present recovery + normal auth release `share_bio` through a robust
/// biometric match against a DEVICE-BOUND sealed secret. This is the ONLY place
/// biometric material is matched (the fuzzy extractor is retired — TRUST_LAYER.md
/// #7; total loss uses the portable shares + the in-person ceremony). The protocol
/// lets `AtlasCore.Recovery` stay pure: the app injects the REAL Secure Enclave
/// (AtlasApp/Enclave), and tests/previews inject `ModelEnclave`.
///
/// Invariant: the biometric is never exposed — the enrolled template is sealed
/// under a non-extractable key and matched inside the boundary.
public protocol BiometricEnclave {
    var deviceID: Data { get }
    /// Whether a biometric is already enrolled/bound on this enclave. Callers must
    /// NOT re-enrol over an existing binding (mirrors Python SecureEnclave.has_biometric):
    /// overwriting the template re-points every secret sealed under it to a new biometric.
    var hasBiometric: Bool { get }
    func enrolBiometric(_ template: Data)
    /// Seal a secret to THIS device, releasable only on a biometric match.
    func seal(_ secret: Data, label: Data) -> Data
    /// Release iff the live biometric matches (robustly) on THIS device; nil
    /// otherwise or if the blob was sealed on a different device.
    func release(_ sealed: Data, liveSample: Data, label: Data) -> Data?
}

/// Pure-Swift model of one device's Enclave (Mac tests / previews). The real
/// device implementation is `SecureEnclaveStore` in AtlasApp, backed by a
/// Secure Enclave key gated by `biometryCurrentSet` + LAContext.
public final class ModelEnclave: BiometricEnclave {
    public let deviceID: Data
    private let master: Data                  // models the non-extractable HW key
    private var sealedTemplate: Data?

    /// Apple's matcher tolerates real variation; "robust" = within this bit-diff.
    public static let robustMatchMaxBitDiff = 0.35

    public init(deviceID: Data = Primitives.randomBytes(16)) {
        self.deviceID = deviceID
        self.master = Primitives.randomBytes(32)
    }

    public var hasBiometric: Bool { sealedTemplate != nil }

    public func enrolBiometric(_ template: Data) {
        sealedTemplate = try? Primitives.aeadEncrypt(key: master, plaintext: template, aad: Data("atlas/se/tmpl".utf8))
    }

    private func match(_ sample: Data) -> Bool {
        guard let sealed = sealedTemplate,
              let template = try? Primitives.aeadDecrypt(key: master, blob: sealed, aad: Data("atlas/se/tmpl".utf8)),
              template.count == sample.count, !template.isEmpty else { return false }
        var diff = 0
        for (a, b) in zip(template, sample) { diff += (a ^ b).nonzeroBitCount }
        return Double(diff) / Double(template.count * 8) <= Self.robustMatchMaxBitDiff
    }

    private func aad(_ label: Data) -> Data { Data("atlas/se/seal|".utf8) + deviceID + Data("|".utf8) + label }

    public func seal(_ secret: Data, label: Data) -> Data {
        (try? Primitives.aeadEncrypt(key: master, plaintext: secret, aad: aad(label))) ?? Data()
    }

    public func release(_ sealed: Data, liveSample: Data, label: Data) -> Data? {
        guard match(liveSample) else { return nil }
        return try? Primitives.aeadDecrypt(key: master, blob: sealed, aad: aad(label))
    }
}

File: backend/atlas/keys/enclave.py
"""Secure Enclave biometric-bound key release (model) — §0.3, §7.

Models Apple's Secure Enclave biometric release for the *device-present*
recovery paths and normal auth:

  * Enclave release: a robust biometric match UNLOCKS a secret SEALED inside a
    specific device's hardware. Robust on real fingers/faces, but DEVICE-BOUND —
    the sealed secret does not survive device loss (so it cannot be the total-loss
    path; total loss rides the portable threshold shares + the in-person ceremony).

This is the ONLY place biometric material is matched: Atlas extracts no key from
raw biometrics (the fuzzy extractor is retired — TRUST_LAYER.md #7). It preserves
the invariant "never store the biometric": the Enclave keeps the enrolled template
sealed under a non-extractable hardware key and matches inside its boundary; it is
never exposed. This Python model makes that boundary explicit so the stratified
recovery model (recovery.py) is testable off-device.

On iOS this is a real Secure Enclave key gated by `biometryCurrentSet` /
`LAContext`; see ios/AtlasApp/Enclave/SecureEnclaveStore.swift.
"""

from __future__ import annotations

from dataclasses import dataclass

from ..crypto.primitives import aead_decrypt, aead_encrypt, random_bytes

# Apple's matcher tolerates real-world variation. We model "robust" as accepting
# samples within this fraction of differing bits from the enrolled template.
ROBUST_MATCH_MAX_BIT_DIFF = 0.35


def _bit_diff_ratio(a: bytes, b: bytes) -> float:
    if len(a) != len(b) or not a:
        return 1.0
    diff = sum(bin(x ^ y).count("1") for x, y in zip(a, b))
    return diff / (len(a) * 8)


class SecureEnclave:
    """Models ONE physical device's Secure Enclave.

    Holds a non-extractable hardware master key and (after enrolment) a sealed
    biometric template. Neither is ever returned. Secrets sealed here are
    device-bound: another device's Enclave cannot release them.
    """

    def __init__(self, device_id: bytes | None = None):
        self.device_id = device_id or random_bytes(16)
        self._master = random_bytes(32)          # non-extractable hardware key
        self._sealed_template: bytes | None = None

    # -- enrolment ----------------------------------------------------------

    def enrol_biometric(self, template: bytes) -> None:
        """Store the enrolled template SEALED under the hardware key. The
        template is never exposed outside the Enclave (matching is internal)."""
        self._sealed_template = aead_encrypt(self._master, template, aad=b"atlas/se/tmpl")

    @property
    def has_biometric(self) -> bool:
        return self._sealed_template is not None

    # -- matching (inside the boundary) -------------------------------------

    def _match(self, sample: bytes) -> bool:
        if self._sealed_template is None:
            return False
        template = aead_decrypt(self._master, self._sealed_template, aad=b"atlas/se/tmpl")
        return _bit_diff_ratio(template, sample) <= ROBUST_MATCH_MAX_BIT_DIFF

    # -- biometric-bound seal / release -------------------------------------

    def seal(self, secret: bytes, *, label: bytes = b"") -> bytes:
        """Seal a secret to THIS device, releasable only on a biometric match."""
        aad = b"atlas/se/seal|" + self.device_id + b"|" + label
        return aead_encrypt(self._master, secret, aad=aad)

    def release(self, sealed: bytes, *, live_sample: bytes, label: bytes = b"") -> bytes | None:
        """Release a sealed secret iff the live biometric matches (robustly) on
        THIS device. Returns None on mismatch or if the blob was sealed elsewhere
        (device-bound: another Enclave's master key won't decrypt it)."""
        if not self._match(live_sample):
            return None
        aad = b"atlas/se/seal|" + self.device_id + b"|" + label
        try:
            return aead_decrypt(self._master, sealed, aad=aad)
        except Exception:
            return None


@dataclass(frozen=True)
class EnclaveSealedShare:
    """A recovery share sealed for device-present release. Device-bound: it is
    stored on (and only usable by) the enrolled device's Enclave."""

    device_id: bytes
    sealed: bytes
